AudioOutput.h

```diff
@@ -52,72 +52,35 @@
-#define SCALE_AUDIO_NEAR(x,bits) (bits > MOZZI_AUDIO_BITS_OPTIMISTIC ? (x) >> (bits - MOZZI_AUDIO_BITS_OPTIMISTIC) : (x) << (MOZZI_AUDIO_BITS_OPTIMISTIC - bits))
-#define CLIP_AUDIO(x) constrain((x), (-(AudioOutputStorage_t) MOZZI_AUDIO_BIAS), (AudioOutputStorage_t) MOZZI_AUDIO_BIAS-1)
+template<typename T> constexpr AudioOutputStorage_t SCALE_AUDIO(T x, byte bits) { return (bits > MOZZI_AUDIO_BITS ? (x) >> (bits - MOZZI_AUDIO_BITS) : (x) << (MOZZI_AUDIO_BITS - bits)); }
+template<typename T> constexpr AudioOutputStorage_t SCALE_AUDIO_NEAR(T x, byte bits) { return (bits > MOZZI_AUDIO_BITS_OPTIMISTIC ? (x) >> (bits - MOZZI_AUDIO_BITS_OPTIMISTIC) : (x) << (MOZZI_AUDIO_BITS_OPTIMISTIC - bits)); }
+template<typename T> constexpr AudioOutputStorage_t CLIP_AUDIO(T x) { return (constrain((x), (-(AudioOutputStorage_t) MOZZI_AUDIO_BIAS), (AudioOutputStorage_t) (MOZZI_AUDIO_BIAS-1))); }
+
+struct MonoOutput;
+struct StereoOutput;
 
 #if MOZZI_IS(MOZZI_AUDIO_CHANNELS, MOZZI_STEREO)
-#define AudioOutput StereoOutput
-#if (STEREO_HACK == true)
-#define AudioOutput_t void
-#else
-#define AudioOutput_t StereoOutput
-#endif
+typedef StereoOutput AudioOutput;
 #else
-/** Representation of an single audio output sample/frame. For mono output, this is really just a single zero-centered int,
- *  but for stereo it's a struct containing two ints.
- *
- *  While it may be tempting (and is possible) to use an int, directly, using AudioOutput_t and the functions AudioOutput::from8Bit(),
- *  AudioOutput::fromNBit(), or AudioOutput::fromAlmostNBit() will allow you to write code that will work across different platforms, even
- *  when those use a different output resolution.
- *
- *  @note The only place where you should be using AudioOutput_t, directly, is in your updateAudio() function signature. It's really just a
- *        dummy used to make the "same" function signature work across different configurations (importantly mono/stereo). It's true value
- *        might be subject to change, and it may even be void. Use either MonoOutput or StereoOutput to represent a piece of audio output.
- */
-#define AudioOutput_t AudioOutputStorage_t
-/** Representation of an single audio output sample/frame. This #define maps to either MonoOutput or StereoOutput, depending on what is configured
+/** Representation of an single audio output sample/frame. This typedef maps to either MonoOutput or StereoOutput, depending on what is configured
  *  in MOZZI_AUDIO_CHANNELS. Since the two are source compatible to a large degree, it often isn't even necessary to test, which it is, in your code. E.g.
  *  both have functions l() and r(), to return "two" audio channels (which will be the same in case of mono).
  *
  *  You will not usually use or encounter this definition, unless using @ref external_audio output mode.
  */
-#define AudioOutput MonoOutput
+typedef MonoOutput AudioOutput;
```

Comment:
MonoOutput is still an `int` internally right?

Reply:
No, but yes ;-). It's a class, but with no virtual functions, and an int as only data member, which means it is stored as an int. It also has implicit conversion to, and from int, so flash/RAM usage stays the same, and everything that used to be possible still is,.

*Except* when MOZZI_COMPATIBILITY_LEVEL is set to greater than 1.1 (which, by default, is the case, when including Mozzi.h, but not when including MozziGuts.h), defining updateAudio() as returning int is no longer accepted.